### PR TITLE
[4.0] Fix the bad implementation for the XTD-buttons on tinyMCE

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue # .

### Summary of Changes
**Without breaking the current API for the XTD-Buttons** this PR is dropping the js that was passed in the json (a very bad practice for which I'll take the blame!). So instead of rendering a list of buttons it renders a dropdown with all the XTD-Buttons as links (with their icons). This can be improved by separating the core xt-buttons from the ones from 3rd PD for clarity and also to have more manageable dropdowns (this is a TODO, if people agree on the idea).

Also since we don't have the `jModalClose()` which we hacked to somehow work with bootstrap and since we are not gonna use the Bootstrap modal in J4 I propose to use the `Joomla` in this way:
```js
Joomla.currentModal = {
  get: function() { 
    // check if nested 
    return this.currentModal;
  },
  set: function(element) {
   // check if nested
  this.currentModal = element;
  },
 currentModal: '',
 parentModal: ''
}
```

##### So how this thing works?
When a modal is called the setter is also called and it, what else, sets the element of the current modal.
Now wherever in the code (especially if you're not aware of the modal element) you can call:
```js
jQuery(Joomla.currentModal.get()).modal('hide');
// or using the J4 UI modal
Joomla.currentModal.get().close();
```
### Preview
screen shot 2018-02-26 at 19 00 50

### Testing Instructions
Apply patch and test all the buttons in tinyMCE

### Documentation Changes Required
NO

@Fedik @dneukirchen what do you think?

### CODE REVIEW

File: layouts/joomla/editors/buttons/button.php [hunk 1]
```diff
@@ -22,7 +22,7 @@
 	$onclick = ($button->get('onclick')) ? ' onclick="' . $button->get('onclick') . '"' : '';
 	$title   = ($button->get('title')) ? $button->get('title') : $button->get('text');
 ?>
-<a href="<?php echo $href; ?>" role="button" class="<?php echo $class; ?>" <?php echo $button->get('modal') ? 'data-toggle="modal"' : '' ?> title="<?php echo $title; ?>" <?php echo $onclick; ?>>
+<a href="<?php echo $href; ?>" role="button" class="btn btn-secondary <?php echo $class; ?>" <?php echo $button->get('modal') ? 'data-toggle="modal"' : '' ?> title="<?php echo $title; ?>" <?php echo $onclick; ?>>
```
Comment (dgrammatiko) [inline layouts/joomla/editors/buttons/button.php:25]: The class belongs to the layout, don't set it in the buttons logic!!!

**Comment by brianteeman** [comment]
Based on the screenshot

What is the ordering of these? It should default to be alphabetical imho

**Comment by dgrammatiko** [comment]
> What is the ordering of these?

I have no clue, but I agree this doesn't look nice

File: plugins/editors/tinymce/tinymce.php [hunk 2]
```diff
@@ -1069,7 +1011,7 @@ public static function getToolbarPreset()
 				'link', 'unlink', 'anchor', 'code', '|',
 				'hr', 'table', '|',
 				'subscript', 'superscript', '|',
-				'charmap', 'pastetext' , 'preview'
+				'charmap', 'pastetext' , 'preview', 'mybutton'
```
Comment (Quy) [inline plugins/editors/tinymce/tinymce.php]: Remove blank space after `'pastetext'`.

**Comment by ggppdk** [comment]
> What is the ordering of these? It should default to be alphabetical imho
> I have no clue, but I agree this doesn't look nice

It seems to be the ordering that plugins have in plugin manager when you filter via editors-xtd type

File: media/editors/tinymce/js/tinymce.js [hunk 3]
```diff
@@ -62,18 +58,39 @@
...
+					};
+				}
+
+				buttonValues.push(tmp)
+			});
+
+			options.setup = function (editor) {
+				editor.addButton('jxtdbuttons', {
+					type   : 'menubutton',
+					text   : 'CMS Content',
```
Comment (brianteeman) [inline media/editors/tinymce/js/tinymce.js]: This should be a translatable string

**Comment by Fedik** [comment]
@dgt41 looks good, 
I just not sure whether it will work in multi instances, eg in subform,
I try to find a time to check

**Comment by dgrammatiko** [comment]
@Fedik it should, the building process relies on the current editor settings/params

**Comment by Fedik** [comment]
I meant the part with modal and it`s id:
```javascript
var modal = document.getElementById(name.name.replace(' ', '') + 'Modal');
```

**Comment by dgrammatiko** [comment]
@Fedik oops those modals are not unique enough, they also need the id of the editor there, easy patch

**Comment by brianteeman** [comment]
They need to be unique for a11y as well

**Comment by Fedik** [comment]
@dgt41 maybe we could use `class` instead, and make more strict selector,
something like (fake code):
```javascript
var editorWrapper = editor.getParentElement();
var modal = editorWrapper.querySelector('.modal-window-media');
```

with this you can just drop the id attribute

**Comment by dgrammatiko** [comment]
@Fedik the editors, not only tinyMCE will eventually become Custom Elements because frankly they are fields... So we can implement a lot more things once we do the transition. Basically this PR was done just as a POC and also to get some feedback on the needed API change for closing the current modal, which is something that we miss in J4 and thus a lot of scripts are failing...

File: media/editors/tinymce/js/tinymce.js (hunk 3 above)
Comment (dgrammatiko) [reply media/editors/tinymce/js/tinymce.js]: done

**Comment by dgrammatiko** [comment]
Ok, multiple instances get modals with unique ids
So:
screen shot 2018-02-27 at 15 55 25
screen shot 2018-02-27 at 15 54 46

**Comment by brianteeman** [comment]
looks good but the icons will need to match the icons in the menu for those components ;)

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 4b1b8e9bc2aa5873b106c077f5aefc10ccd8aea2<br /><br />Notice: Undefined index: native in C:\htdocs\cms4\plugins\editors\tinymce\tinymce.php on line 531

Warning: array_merge(): Argument #2 is not an array in C:\htdocs\cms4\plugins\editors\tinymce\tinymce.php on line 531

Notice: Undefined index: paths in C:\htdocs\cms4\plugins\editors\tinymce\tinymce.php on line 534

Warning: array_merge(): Argument #2 is not an array in C:\htdocs\cms4\plugins\editors\tinymce\tinymce.php on line 534
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19789">issues.joomla.org/tracker/joomla-cms/19789</a>.</sub>

**Comment by dgrammatiko** [comment]
@brianteeman thanks, now should be fine

PS about the icons:
This is the before
screen shot 2018-02-27 at 20 37 03

this is the after
screen shot 2018-02-27 at 15 55 25

So in essence those icons are not changed. Maybe in the new skin that is proposed by the design team we can use the font awesome (or a part of it, don't forget that tiny is still in an iframe) to get some more meaningful icons

File: plugins/editors/tinymce/tinymce.php [hunk 4]
```diff
@@ -792,14 +734,14 @@ protected static function getGlobalFilters()
 
 		$filters = $config->get('filters');
 
-		$blackListTags       = array();
-		$blackListAttributes = array();
+		$blackListTags        = array();
```
Comment (Quy) [inline plugins/editors/tinymce/tinymce.php:737]: Change to `[]`

File: plugins/editors/tinymce/tinymce.php [hunk 5]
```diff
@@ -1044,18 +986,18 @@ public static function getKnownButtons()
...
-				'pastetext'
-			),
-			'toolbar2' => array(),
-		);
+				'pastetext', 'jxtdbuttons'
+			],
+			'toolbar2' => [],
+		];
 
 		$preset['medium'] = array(
```
Comment (Quy) [inline plugins/editors/tinymce/tinymce.php:1002]: Change to `[]`

File: plugins/editors/tinymce/tinymce.php [hunk 6]
```diff
@@ -525,30 +525,24 @@ public function onDisplay(
...
-		$toolbar2  = array_merge($toolbar2, $btns['native']);
-
-		// Merge the custom plugins paths
-		$externalPlugins  = array_merge($externalPlugins, $btns['paths']);
-
 		// Build the final options set
-		$scriptOptions = array_merge(
+		$scriptOptions   = array_merge(
 			$scriptOptions,
 			array(
```
Comment (Quy) [inline plugins/editors/tinymce/tinymce.php:533]: Change to `[]`

Comment (dgrammatiko) [reply plugins/editors/tinymce/tinymce.php:533]: @Quy this whole file need a good refactor, let's do that in another PR, touching too many lines makes it harder for people to review the essential changes

File: plugins/editors-xtd/image/image.php [hunk 7]
```diff
@@ -36,7 +36,7 @@ class PlgButtonImage extends CMSPlugin
 	 * @param   string   $asset   The name of the asset being edited.
 	 * @param   integer  $author  The id of the author owning the asset being edited.
 	 *
-	 * @return  JObject  The button options as JObject or false if not allowed
+	 * @return  mixed (CMSObject|| false)  The button options as JObject or false if not allowed
```
Comment (wilsonge) [inline plugins/editors-xtd/image/image.php]: `CMSObject|false` instead of mixed

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 3c21b0738cbcee33e57f43e3f3204caf659d048d<br /><br />no xtd-buttons at all now<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19789">issues.joomla.org/tracker/joomla-cms/19789</a>.</sub>

**Comment by dgrammatiko** [comment]
@brianteeman please clear your browser's cache

**Comment by brianteeman** [comment]
@dgt41 I am not a newbie - of course I cleared the cache. I even tested it on multiple browsers - at least one of which has never been used before on this computer for anything at all (ie11)

**Comment by dgrammatiko** [comment]
@brianteeman just curious here, did you used patchtester for this one?

**Comment by brianteeman** [comment]
Probably I don't remember

**Comment by brianteeman** [comment]
now that I have escaped scotland I will try again with git this week

**Comment by brianteeman** [comment]
must have been a patchtester issue as installing your fork works correctly as expected

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 9dbfcb5fb9c63a75f813a50afcde99a36a70b009<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19789">issues.joomla.org/tracker/joomla-cms/19789</a>.</sub>

**Comment by anuragteapot** [comment]
I have tested this item :white_check_mark: successfully on 9dbfcb5fb9c63a75f813a50afcde99a36a70b009<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19789">issues.joomla.org/tracker/joomla-cms/19789</a>.</sub>

**Comment by dgrammatiko** [comment]
@wilsonge if you agree with this one can you please merge it?

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19789">issues.joomla.org/tracker/joomla-cms/19789</a>.</sub>

**Comment by dgrammatiko** [comment]
@regularlabs can you have a look on this one and provide some feedback?

**Comment by regularlabs** [comment]
I haven't tested this. But I can give you my opinion.

Placing all these editor button plugins in a separate dropdown feels a bit weird to me.
I would leave this to be controlled by the administrator. Let him/her decide if stuff needs to be grouped, and which buttons should get grouped.

Having stuff you use a lot and want there at-the-ready - like the read more button - is now hidden away in some undescriptive bucket full of random stuff that most likely have nothing to do with each other.
Also there are editor buttons that will show a dropdown/submenu thing themselves (at least one, Content Templater ;) ). So that will pose an issue for that extension.

Personally I was not very happy with the editor buttons getting forced into the editor in the first place. It caused all sorts of issues (styling, scripts, etc). I expect the same thing with this new approach.
I still prefer just using JCE (like everyone else) which shows the editor buttons under the editor (in Joomla 3 anyway), like we were use to.

But I am not very bothered about any of it.
As an extension developer I will just work my way around any new issues or workflows Joomla throws at me.

**Pull request merged**